supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala

```diff
@@ -0,0 +1,146 @@
+      jsonStringIo.map(_.parseJson.convertTo[ManifestFile])
+    }
+  }
+
+  protected def bulkValidateCrc32cs(referenceFiles: Set[ReferenceFile]): IO[Map[ReferenceFile, Boolean]] = {
+    val gcsClient = createGcsClient()
+
+    val filesAndValidatedPaths = referenceFiles.map {
+      referenceFile => (referenceFile, GcsPathBuilder.validateGcsPath(s"gs://${referenceFile.path}"))
+    }.toMap
+
+    val filesWithValidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, validPath: ValidFullGcsPath) => (referenceFile, validPath)
+    }
+    val filesWithInvalidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, invalidPath: InvalidFullGcsPath) => (referenceFile, invalidPath)
+    }
+
+    if (filesWithInvalidPaths.nonEmpty) {
+      IO.raiseError(new RuntimeException(s"Some of the paths in manifest file are not valid GCS paths: " +
+        s"${filesWithInvalidPaths.keySet.map(_.path).toList.mkString(", ")}")
+      )
+    } else {
+      IO {
+        val gcsBatch = gcsClient.batch()
+        val filesAndBlobResults = filesWithValidPaths.map {
+          case (referenceFile, ValidFullGcsPath(bucket, path)) =>
+            val blobGetResult = gcsBatch.get(BlobId.of(bucket, path.substring(1)), BlobGetOption.fields(BlobField.CRC32C))
+            (referenceFile, blobGetResult)
+        }
+        gcsBatch.submit()
+
+        filesAndBlobResults.map {
+          case (referenceFile, blobGetResult) =>
+            val crc32cFromManifest = BaseEncoding.base64.encode(
+              // drop 4 leading bytes from Long crc32c value
+              // https://stackoverflow.com/a/25111119/1794750
+              util.Arrays.copyOfRange(Longs.toByteArray(referenceFile.crc32c), 4, 8)
+            )
+
+            (referenceFile, crc32cFromManifest === blobGetResult.get().getCrc32c)
+        }
+      }
+    }
+  }
+
+  private def manifestFileToMapOfValidReferences(manifestFile: ManifestFile): IO[Map[String, PipelinesApiReferenceFilesDisk]] = {
+    val refDisk = PipelinesApiReferenceFilesDisk(manifestFile.imageIdentifier, manifestFile.diskSizeGb)
+    val allReferenceFilesFromManifestMap = manifestFile.files.map(refFile => (refFile, refDisk)).toMap
+    val filesWithValidatedCrc32csIo = bulkValidateCrc32cs(allReferenceFilesFromManifestMap.keySet)
+
+    val validReferenceFilesFromManifestMapIo = filesWithValidatedCrc32csIo map {
+      filesWithValidatedCrc32c => {
+        allReferenceFilesFromManifestMap.filterKeys(key => filesWithValidatedCrc32c.getOrElse(key, false))
+      }
+    }
+
+    validReferenceFilesFromManifestMapIo map { validReferenceFilesFromManifestMap =>
+      val invalidReferenceFiles = allReferenceFilesFromManifestMap.keySet -- validReferenceFilesFromManifestMap.keySet
+      if (invalidReferenceFiles.nonEmpty) {
+        logger.warn(s"The following files listed in references manifest have checksum mismatch with actual files in GCS: ${invalidReferenceFiles.mkString(",")}")
```

Comment:
If I'm reading this right, it looks really nice - so if a reference is no longer valid, we get a warning about it, but it doesn't stop us from using the rest of the (still-valid) references on the disk?

Reply:
Yes, this is correct